src/compiler/scala/tools/nsc/ast/parser/Parsers.scala

```diff
@@ -1635,9 +1635,19 @@ self =>
             case Parens(xs) if xs.forall(isTypedParam) => true
             case _ => false
           }
-          if (in.token == ARROW && (location != InTemplate || lhsIsTypedParamList)) {
-            t = atPos(t.pos.start, in.skipToken()) {
-              Function(convertToParams(t), if (location != InBlock) expr() else block())
+
+          def lhsIsSelfType = t match {
+            case Parens(tpeds) => tpeds.collectFirst{case Typed(This(_), _) => ()}.isDefined
+            case _ => false
+          }
+
+          if (in.token == ARROW){
+            if (lhsIsSelfType) {
+              reporter.error(t.pos, "self-type annotation may not be in parentheses")
+            } else if (location != InTemplate || lhsIsTypedParamList) {
+              t = atPos(t.pos.start, in.skipToken()) {
+                Function(convertToParams(t), if (location != InBlock) expr() else block())
+              }
```

Comment:
Actually,
```
          def lhsIsTypedParamList() = t match {
            case Parens(Typed(This(_), _) :: Nil) => reporter.error(t.pos, "self-type annotation may not be in parentheses") ; false
            case Parens(xs)                       => xs.forall(isTypedParam)
            case _                                => false
          }
```
because you only want to snag what looks like a self-type
```
scala> class C { (this: Int) => 42 }
                 ^
       error: self-type annotation may not be in parentheses

scala> class C { (this: Int, x: Int) => 42 }
                      ^
       error: not a legal formal parameter.
       Note: Tuples cannot be directly destructured in method or function parameters.
             Either create a single parameter accepting the Tuple1,
             or consider a pattern matching anonymous function: `{ case (param1, param1) => ... }
```
The call from `templateStatSeq` for the self-type should really specify "don't strip parens" so it can do the specific message, and elsewhere in the template it could get the ordinary message, but it doesn't matter much.

Reply:
Yes, that makes sense!